runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -151,45 +178,81 @@ public int resolve(
         }
         else
         {
+            CachedSchemaId cachedSchemaId = null;
+            AtomicInteger eventStatus = new AtomicInteger();
+            long retry = RESET_RETRY_DELAY_MS_DEFAULT;
             CompletableFuture<CachedSchemaId> newFuture = new CompletableFuture<>();
+            CompletableFuture<CachedSchemaId> existing = cachedSchemaIds.get(schemaKey);
+            if (existing != null && existing.isDone())
+            {
+                try
+                {
+                    cachedSchemaId = existing.get();
+                    if (cachedSchemaId != null)
+                    {
+                        eventStatus = cachedSchemaId.event;
+                    }
+                }
+                catch (Throwable ex)
+                {
+                    existing.completeExceptionally(ex);
+                }
+            }
             CompletableFuture<CachedSchemaId> future = cachedSchemaIds.merge(schemaKey, newFuture, (v1, v2) ->
-                v1.getNow(IN_PROGRESS).expired(maxAgeMillis) ? v2 : v1);
+                v1.getNow(IN_PROGRESS).retry() &&
+                    (v1.getNow(IN_PROGRESS).id == NO_SCHEMA_ID || v1.getNow(IN_PROGRESS).expired(maxAgeMillis)) ? v2 : v1);
             if (future == newFuture)
             {
                 try
                 {
                     String response = sendHttpRequest(MessageFormat.format(SUBJECT_VERSION_PATH, subject, version));
+                    if (response == null)
+                    {
+                        if (eventStatus.getAndIncrement() == 0)
+                        {
+                            retry = RETRY_INITIAL_DELAY_MS_DEFAULT;
+                            event.unretrievableSchemaSubjectVersion(catalogId, subject, version);
+                        }
+
+                        if (cachedSchemaId != null && cachedSchemaId.retry != RESET_RETRY_DELAY_MS_DEFAULT)
+                        {
+                            long current = cachedSchemaId.retry;
+                            long update = current * RETRY_MULTIPLER_DEFAULT;
+                            retry = update < maxAgeMillis ? update : current;
```

Comment:
This logic is trying to prevent the value from increasing if it would end up larger than `maxAgeMillis`, but that means `maxAgeMillis` is okay too.

Suggest it is easier to follow if you change to max out at `maxAgeMillis` instead of `current`.

```suggestion
retryAfter = Math.min(cachedSchemaId.retryAfter << 1, maxAgeMillis);
```